### PR TITLE
TechDocs: Access entity techdocs from service catalog

### LINKED ISSUE #1810: TechDocs: Figure out when to build documentation for a service. /buildall endpoint won't scale well.

### LINKED ISSUE #1811: TechDocs: Add links to Documentation from entities that have a documentation reference in service catalog.

### PR DESCRIPTION
**Pull request by hooloovooo**
Adds techdocs tab to service catalogs entity page.

Closes #1811 #1810 

Screenshot from 2020-08-06 15-41-25

### CODE REVIEW

**Comment by stefanalund** [comment]
I think the tab should simply say DOCS.

**Comment by hooloovooo** [comment]
Not sure what the Frontend CI test is complaining about. Don't get that error locally and adding the package causes errors because it'll have double typedefs.

**Comment by hooloovooo** [comment]
This will also change the URL structure required by the docs host to be <storageUrl>/<entity kind>/<namespace>/<name> instead of just <storageUrl>/<name>. Is this something we will have to update somewhere? @stefanalund

**Comment by freben** [comment]
@hooloovooo you need to quote that with backticks, it got lost

**Comment by freben** [comment]
@hooloovooo i took the liberty of rebasing this and fixing up yarn.lock because it had changes in it that i think you forgot to add, let's see if it builds now

**Comment by andrewthauer** [comment]
> This will also change the URL structure required by the docs host to be \<storageUrl\>/\<entity kind\>/\<namespace\>/\<name\> instead of just \<storageUrl\>/\<name\>. Is this something we will have to update somewhere? @stefanalund 

Is this URL convention going to be required or can it be configured? As it stands, we can use the techdocs reader as is with the simple convention of storageBase/name. Making this change mandatory will require various changes to accommodate.

File: plugins/techdocs/src/plugin.ts
Comment (andrewthauer) [inline plugins/techdocs/src/plugin.ts:42]: 😀 Would it be possible to make this overrideable via an apiRef interface? That way we can still use this with our pre-existing docs storage & build setup.

File: plugins/techdocs/src/reader/components/Reader.tsx
Comment (andrewthauer) [inline plugins/techdocs/src/reader/components/Reader.tsx:89]: 😀 Would it be possible to make this overrideable via an apiRef interface? That way we can still use this with our pre-existing docs storage & build setup.

Comment (hooloovooo) [reply plugins/techdocs/src/reader/components/Reader.tsx:89]: Yea that's definitely something we should add. Just having storageBase/name could cause issues though since the name is not a unique identifier. I'd recommend you either store by ID or this combination of kind, namespace and name.

Comment (andrewthauer) [reply plugins/techdocs/src/reader/components/Reader.tsx:89]: Thanks! In our case the name is the repo name, so it will be unique. We have pre-exiting CI steps to publish to our storage location based on the repo name. Having the ability to override this will avoid us having to use the existing convention with minimal changes.

File: plugins/techdocs/src/api.ts
Comment (andrewthauer) [inline plugins/techdocs/src/api.ts:25]: ❤️ Thank you @hooloovooo!

**Comment by hooloovooo** [comment]
Could we get a quick review from @spotify/backstage-core on this? There's still more work to do here, but it's growing too so would be nice to merge.

File: plugins/techdocs/src/types.ts
Comment (freben) [inline plugins/techdocs/src/types.ts:19]: usually specified as optional `?`

File: plugins/techdocs/src/api.ts
Comment (freben) [inline plugins/techdocs/src/api.ts:45]: how does this work out when there is no namespace? do you get `default` at this point?

File: plugins/catalog/src/routes.ts
Comment (freben) [inline plugins/catalog/src/routes.ts:28]: hmm how does this work when there is no selected tab ID? like, doesn't this path REQUIRE that all of those things are specified

File: plugins/catalog/src/components/EntityPage/EntityPage.tsx
Comment (freben) [inline plugins/catalog/src/components/EntityPage/EntityPage.tsx:241]: i'm not sure about this data passing - how about

```
  const tabs = [
    {
      id: 'overview',
      label: 'Overview',
      content: () => <EntityOverviewPage entity={entity} />
    }
```

That way we get to break out all of those page contents elsewhere and don't have to invent a custom passing mechanism. And then we could just render the `HeaderTabs` directly in here and handle its onChange with navigation in the same place as the other knowledge in here

File: plugins/techdocs/src/reader/components/TechDocsHome.tsx
Comment (freben) [inline plugins/techdocs/src/reader/components/TechDocsHome.tsx:72]: @soapraj That entity ID RFC will come in handy :)

Comment (Rugvip) [reply plugins/techdocs/src/reader/components/TechDocsHome.tsx:72]: Yup :p

File: plugins/catalog/src/routes.ts
Comment (hooloovooo) [reply plugins/catalog/src/routes.ts:28]: Yup, it does. I've changed it to point to /overview where i found it linking to the entity page. Would be a good thing to have a default as well.

File: plugins/techdocs/src/api.ts
Comment (hooloovooo) [reply plugins/techdocs/src/api.ts:45]: Yes, i make sure it's set to default before being passed there. Could probably change that to happen in here instead since it should probably only be something the API has to know about.

**Comment by hooloovooo** [comment]
@spotify/backstage-core I have revisited the things that have comments and it should be in better shape now.

File: packages/core/src/layout/HeaderTabs/index.tsx
Comment (freben) [inline packages/core/src/layout/HeaderTabs/index.tsx:51]: Do these tabs offer the ability to BOTH support active and passive mode so to speak? I was glad to see the `selectedIndex` prop, and I think that if we can get away with it, we should use that strategy everywhere (forcing the caller to manage the tab state). Does that make sense? Then we would remove this `selectedTab` state entirely and simplify this component a bit.

I won't block this PR on that, but it should probably be the "way of the future" :)

**Review by freben — APPROVED** [review]

File: packages/core/src/layout/HeaderTabs/index.tsx
Comment (hooloovooo) [reply packages/core/src/layout/HeaderTabs/index.tsx:51]: Yea, this makes it support both. Settling on one way would be nice, but i'll leave this here for now. No idea what would break if i removed it's own state management.

**Pull request merged**
